### PR TITLE
blog: upgrade td-agent v4 to fluent-package v5

### PR DESCRIPTION
**Pull request by kenhys**

### CODE REVIEW

**Comment by kenhys** [comment]
If release target day is determined, it should update the file name of blog entry.

File: content/blog/20230731_upgrade-td-agent-v4-to-v5.md [hunk 1]
```diff
@@ -0,0 +1,286 @@
...
+There are two channels for `fluent-package`.
+
+* Standard version
+* Long Term Support version
+
+One is standard version which will be updated regularly - rapid release edition. (`td-agent v4` was released like this in almost every quarter)
+In this version, fluentd will be eventually updated to newer minor version (e.g. 1.17.x and so on)
+
+The other is a more conservative maintenance version (Long term support) which will not introduce a new feature. It only applies teeny update such as
+security or bug fix only. LTS version will be supported until March, 2025.
```
Comment (daipom) [inline content/blog/20230731_upgrade-td-agent-v4-to-v5.md:28]:
```suggestion
security or bug fix only. LTS for v5 will be supported until March, 2025.
```

I think we should make it clear that this period is about v5.

File: content/blog/20230731_upgrade-td-agent-v4-to-v5.md [hunk 2]
```diff
@@ -0,0 +1,286 @@
...
+drwxr-xr-x.  3 root root   49  7月 14 03:14 fluent-plugin-systemd-1.0.5
+drwxr-xr-x.  5 root root  221  7月 14 03:14 fluent-plugin-td-1.2.0
+drwxr-xr-x.  5 root root  166  7月 14 03:14 fluent-plugin-utmpx-0.5.0
+drwxr-xr-x.  4 root root  191  7月 14 03:14 fluent-plugin-webhdfs-1.5.0
+```
+
+### 8. Start fluent-package v5 daemon.
+
+```
+$ sudo systemctl start fluent-package
```
Comment (daipom) [inline content/blog/20230731_upgrade-td-agent-v4-to-v5.md:235]:
```suggestion
$ sudo systemctl start fluentd
```

File: content/blog/20230731_upgrade-td-agent-v4-to-v5.md [hunk 3]
```diff
@@ -0,0 +1,286 @@
...
+Now, upgrading steps are completed. Happy Logging!
+
+## Additional hints for v4 users
+
+### For Debian/Ubuntu
+
+* If you upgrade from v4, `td-agent` will be installed as a transitional package.
+  you can remove it safely with `sudo apt purge td-agent`.
+* `fluentd-apt-source` package will be marked as a transitional package.
+  you can remove it safely with `sudo apt purge fluentd-apt-source`.
```
Comment (daipom) [inline content/blog/20230731_upgrade-td-agent-v4-to-v5.md:253]: It looks like this says that users who upgraded from v4 need to uninstall v4.
Is this intentional?

Or do you want to say that users can uninstall v5 safely even if it is upgraded from v4?
Then, is `sudo apt purge fluent-package` correct?

File: content/blog/20230731_upgrade-td-agent-v4-to-v5.md [hunk 4]
```diff
@@ -0,0 +1,286 @@
...
+
+  ```
+  $ sudo systemctl unmask td-agent
+  $ sudo systemctl enable fluentd
+  ```
+
+### For RHEL
+
+* If you upgrade from v4, `td-agent` will be replaced with `fluent-package`.
+  you can remove it safely with `sudo dnf remove td-agent`.
```
Comment (daipom) [inline content/blog/20230731_upgrade-td-agent-v4-to-v5.md:264]: Same concern as above.

**Review by daipom — COMMENTED** [review]
Thanks for summarizing this!
I have commented on some points.

File: content/blog/20230731_upgrade-td-agent-v4-to-v5.md (hunk 3 above)
Comment (kenhys) [reply content/blog/20230731_upgrade-td-agent-v4-to-v5.md:253]: No need to uninstall in practically (it is ok to keep it), but transitional package (td-agent) is usually empty, so added note for it.

Comment (daipom) [reply content/blog/20230731_upgrade-td-agent-v4-to-v5.md:262]: Hmm, I think the existing v4 package is automatically uninstalled when upgrading.

I have confirmed the behavoir on RockyLinux 8.
I have upgraded from v4 to v5.
Then, I couldn't find `td-agent`.

```console
[vagrant@localhost ~]$ dnf list --installed | grep flu
Failed to set locale, defaulting to C.UTF-8
fluent-package.x86_64               5.0.0-1.el8                            @@commandline
[vagrant@localhost ~]$ dnf list --installed | grep td
Failed to set locale, defaulting to C.UTF-8
libstdc++.x86_64                    8.5.0-18.el8                           @anaconda    
libtdb.x86_64                       1.4.7-1.el8                            @anaconda    
libzstd.x86_64                      1.4.4-1.el8                            @anaconda
```

When I uninstalled `td-agent`, `fluent-package` is uninstalled.
(I didn't expect this... Looks like `td-agent` is an alias for `fluent-package`...)

```console
[vagrant@localhost ~]$ sudo dnf remove td-agent
Failed to set locale, defaulting to C.UTF-8
Dependencies resolved.
==============================================================================================================================================================================================================
 Package                                             Architecture                                Version                                             Repository                                          Size
==============================================================================================================================================================================================================
Removing:
 fluent-package                                      x86_64                                      5.0.0-1.el8                                         @@commandline                                       64 M

Transaction Summary
==============================================================================================================================================================================================================
Remove  1 Package

Freed space: 64 M
Is this ok [y/N]: y
Running transaction check
Transaction check succeeded.
Running transaction test
Transaction test succeeded.
Running transaction
  Preparing        :                                                                                                                                                                                      1/1 
  Running scriptlet: fluent-package-5.0.0-1.el8.x86_64                                                                                                                                                    1/1 
  Erasing          : fluent-package-5.0.0-1.el8.x86_64                                                                                                                                                    1/1 
warning: /etc/fluent/fluentd.conf saved as /etc/fluent/fluentd.conf.rpmsave

  Running scriptlet: fluent-package-5.0.0-1.el8.x86_64                                                                                                                                                    1/1 
Removing fluentd user...
userdel: group fluentd is the primary group of another user and is not removed.
Removing fluentd group...
groupdel: cannot remove the primary group of user 'td-agent'
Removing td-agent user...
userdel: td-agent mail spool (/var/spool/mail/td-agent) not found
userdel: td-agent home directory (/var/lib/td-agent) not found

  Verifying        : fluent-package-5.0.0-1.el8.x86_64                                                                                                                                                    1/1 

Removed:
  fluent-package-5.0.0-1.el8.x86_64                                                                                                                                                                           

Complete!
[vagrant@localhost ~]$ dnf list --installed | grep flu
Failed to set locale, defaulting to C.UTF-8
[vagrant@localhost ~]$ fluentd --version
-bash: fluentd: command not found
[vagrant@localhost ~]$ systemctl status fluentd
Unit fluentd.service could not be found.
[vagrant@localhost ~]$ systemctl status td-agent
Unit td-agent.service could not be found.
```

File: content/blog/20230731_upgrade-td-agent-v4-to-v5.md (hunk 4 above)
Comment (kenhys) [reply content/blog/20230731_upgrade-td-agent-v4-to-v5.md:273]: Oops, it is just wrong description. I'll fix it.

Comment (kenhys) [reply content/blog/20230731_upgrade-td-agent-v4-to-v5.md]: Fixed.

File: content/blog/20230731_upgrade-td-agent-v4-to-v5.md (hunk 3 above)
Comment (daipom) [reply content/blog/20230731_upgrade-td-agent-v4-to-v5.md:262]: Sorry, maybe I should confirm this on Ubuntu/Debian.
I'm confirming it now.

Comment (daipom) [reply content/blog/20230731_upgrade-td-agent-v4-to-v5.md:262]: I have confirmed this on Ubuntu Focal.
As this document says, there seems to be the `td-agent` package remaining, but I couldn't purge it.
It seems that we need to uninstall v5 first to uninstall the remaining `td-agent` package.

```console
vagrant@vagrant:~$ apt list --installed | grep fluen

WARNING: apt does not have a stable CLI interface. Use with caution in scripts.

fluent-package/now 5.0.0-1 amd64 [installed,local]
fluentd-apt-source/focal,now 2020.8.25-1 all [installed]
vagrant@vagrant:~$ apt list --installed | grep td-agent

WARNING: apt does not have a stable CLI interface. Use with caution in scripts.

vagrant@vagrant:~$ sudo apt purge td-agent
Reading package lists... Done
Building dependency tree       
Reading state information... Done
The following packages will be REMOVED:
  td-agent*
0 upgraded, 0 newly installed, 1 to remove and 194 not upgraded.
After this operation, 0 B of additional disk space will be used.
Do you want to continue? [Y/n] 
(Reading database ... 50480 files and directories currently installed.)
Purging configuration files for td-agent (4.5.0-1) ...
td-agent:x:113:117::/var/lib/td-agent:/usr/sbin/nologin
userdel: user td-agent is currently used by process 21510
dpkg: error processing package td-agent (--purge):
 installed td-agent package post-removal script subprocess returned error exit status 8
Errors were encountered while processing:
 td-agent
E: Sub-process /usr/bin/dpkg returned an error code (1)
```

Comment (kenhys) [reply content/blog/20230731_upgrade-td-agent-v4-to-v5.md:262]: Hmm, it is not unexpected...

Comment (kenhys) [reply content/blog/20230731_upgrade-td-agent-v4-to-v5.md:262]: Created as an issue to track.

https://github.com/fluent/fluent-package-builder/issues/564

Comment (kenhys) [reply content/blog/20230731_upgrade-td-agent-v4-to-v5.md:262]: non- fluent-package issue.

Comment (daipom) [reply content/blog/20230731_upgrade-td-agent-v4-to-v5.md:262]: There are still concerns: https://github.com/fluent/fluent-package-builder/issues/564#issuecomment-1669220651

I suppose we can not purge v4 on the environment where the package is upgraded to v5 from v4.

Comment (kenhys) [reply content/blog/20230731_upgrade-td-agent-v4-to-v5.md]: Thanks, rebased and  dropped guide to purge transitional package.

**Review by daipom — APPROVED** [review]
LGTM. Thanks!
When should we merge this?

**Comment by kenhys** [comment]
> When should we merge this?

When GA is ready to announce. (so tweak filename of this article later)

**Comment by kenhys** [comment]
FIXME: replace "standard" phrase

**Comment by kenhys** [comment]
* Replace "standard version" expression to use normal release
* Fixed upgraded OpenSSL version
* Added link to https://www.fluentd.org/blog/fluent-package-scheduled-lifecycle

**Pull request merged**